tests/core/data_layer/test_data_store.py

```diff
@@ -386,10 +386,11 @@ async def test_batch_update(data_store: DataStore, tree_id: bytes32, use_optimiz
         random.seed(100, version=2)
 
         batch: List[Dict[str, Any]] = []
-        keys: List[bytes] = []
+        keys_values: Dict[bytes, bytes] = {}
         hint_keys_values: Dict[bytes, bytes] = {}
         for operation in range(num_batches * num_ops_per_batch):
-            if random.randint(0, 4) > 0 or len(keys) == 0:
+            op_type = random.randint(0, 5)
```

Comment:
Maybe something like `[op_type] = random.choices(["insert", "upsert", "delete"], weights=[6, 2, 2], k=1)`?  (or whatever weights you intend)

Though in some way we need to test both insert and update cases of upsert.  So maybe `["insert", "upsert-insert", "upsert-update", "delete"]`.